### PR TITLE
CATROID-958 Textblock detection x and y sensors return the wrong value

### PR DESCRIPTION
**Pull request by ritiksp2411**
Changed the value of the sensor variable X to Y and Y to -X.

https://jira.catrob.at/browse/CATROID-958

I tested the work on the Honor 9i android device.
 
Signed-off-by: ritiksp2411 <ritik.rjt@gmail.com>

### Your checklist for this pull request
Please review the [contributing guidelines](https://github.com/Catrobat/Catroid/blob/develop/README.md) and [wiki pages](https://github.com/Catrobat/Catroid/wiki/) of this repository.

- [x] Include the name of the Jira ticket in the PR’s title
- [x] Include a summary of the changes plus the relevant context
- [ ] Choose the proper base branch (*develop*)
- [x] Confirm that the changes follow the project’s coding guidelines
- [x] Verify that the changes generate no compiler or linter warnings
- [x] Perform a self-review of the changes
- [x] Verify to commit no other files than the intentionally changed ones
- [ ] Include reasonable and readable tests verifying the added or changed behavior
- [x] Confirm that new and existing unit tests pass locally
- [x] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Commit-Message-Guidelines)
- [x] Stick to the project’s gitflow workflow
- [x] Verify that your changes do not have any conflicts with the base branch
- [ ] After the PR, verify that all CI checks have passed
- [ ] Post a message in the *catroid-stage* or *catroid-ide* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

### CODE REVIEW

**Comment by qlin93** [comment]
Thanks for making the changes. 
I tested your changes with this test:
testTextblockDetection
and the returned coordinates are not on the screen. The X should be exactly where the text is that is being detected. 
Have a look at FaceAndTextDetector.kt, where the coordinate translation is implemented correctly. You also need to check how it works when using the front camera. Please make the necessary changes and amend your commit.

**Comment by ritiksp2411** [comment]
Thanks for the review, 
I have one doubt like we have a bug in landscape mode right, but the screenshot is in Portrait mode? I didn't get that part.

**Comment by ritiksp2411** [comment]
In the file FaceAndTextDetector.kt, we have Y as X if the camera is facing front, and Y as 1-X if not.
Along with that if we have oldAPI, then X is Y otherwise, X is 1-Y. So do I need to check oldAPI too right?

**Comment by qlin93** [comment]
Thanks. Creating a project in landscape mode does not change the orientation of the Pocket Code app itself only of the project when it is run.

Yes please also take the oldAPI into account.  
Have you installed the app via Android Studio to test the changes you made? 
This is important because the solution for TextblockUtil might not be exactly the same as for the FaceAndTextDetector, 
as the x and y might be different. (It could have different origins)

**Comment by ritiksp2411** [comment]
thanks, I check the changes in my phone but couldn't recognize that properly, now I will make the changes and update accordingly.

**Comment by ritiksp2411** [comment]
When I tested with your test, I have noticed that returned coordinates of the X are sometimes on the screen and sometimes off the screen. Any specific reason for that?

**Comment by ritiksp2411** [comment]
WhatsApp Image 2021-03-25 at 7 45 36 PM
It is showing the X properly now.

**Comment by qlin93** [comment]
Looks good. Please push your changes on this branch by amending your commit.

**Comment by ritiksp2411** [comment]
Thanks for helping, I updated the PR.

**Review by qlin93 — CHANGES_REQUESTED** [review]
Hi sorry for the delay. Unfortunately it still does not work correctly. When a text is detected and you move your camera around the x and y coordinates or in the example testcase the X should follow the text. With the change you made the problem still persists, the movement along the x-axis is shown on the y-axis and vice versa.

**Comment by ritiksp2411** [comment]
okay, yeah so the shift of X was shown in Y, but bow I Fixed it such that when we move our camera left side that means Text in the image will go right and the same will be shown by X on the screen. Similarly, if we move the camera upwards, the text in the image will go down and the same will be shown by X on the screen.

File: catroid/src/main/java/org/catrobat/catroid/utils/TextBlockUtil.java [hunk 1]
```diff
@@ -102,10 +102,10 @@ public static Point getCenterCoordinates(int index) {
 				float aspectRatio = (float) imageWidth / imageHeight;
 
 				if (ProjectManager.getInstance().isCurrentProjectLandscapeMode()) {
-					float relativeY = textBlockBounds.exactCenterX() / imageWidth;
-					relativeY = invertAxis ? relativeY : 1 - relativeY;
+					float relativeY = textBlockBounds.exactCenterY() / imageWidth;
+					relativeY = invertAxis ? 1 - relativeY : relativeY;
```
Comment (qlin93) [inline catroid/src/main/java/org/catrobat/catroid/utils/TextBlockUtil.java:106]:
```suggestion
					float relativeX = textBlockBounds.exactCenterX() / imageWidth;
					relativeX = invertAxis ? 1 - relativeX : relativeX;
```

File: catroid/src/main/java/org/catrobat/catroid/utils/TextBlockUtil.java [hunk 2]
```diff
@@ -102,10 +102,10 @@ public static Point getCenterCoordinates(int index) {
...
 					return coordinatesFromRelativePosition(
-							1 - textBlockBounds.exactCenterY() / imageHeight,
+							1 - textBlockBounds.exactCenterX() / imageHeight,
 							SCREEN_WIDTH / aspectRatio,
 							relativeY,
 							(float) SCREEN_WIDTH
```
Comment (qlin93) [inline catroid/src/main/java/org/catrobat/catroid/utils/TextBlockUtil.java:111]:
```suggestion
							relativeX,
							SCREEN_WIDTH,
							1 - textBlockBounds.exactCenterY() / imageHeight,
							(float) SCREEN_WIDTH / aspectRatio
```

**Review by qlin93 — CHANGES_REQUESTED** [review]
Thanks, that is better but the movement on the axes are still inverted and the scaling is off. I looked into the issue more closely and it turned out to be a little more complicated to fix than stated in the ticket. 

Please apply the suggested changes and test whether they work on your device before updating your commit.

**Comment by ritiksp2411** [comment]
Thanks for the suggestion, I saw that it exactly behaves the same as my previous PR, like axis are behaving the same way. I am unable to justify scaling there but ya, the axis are working in the same manner.

**Comment by qlin93** [comment]
I tested it again and you are right the axes are behaving the same way, but the scaling is completely off. I cannot approve something that does not work correctly and you should always test and make sure that the behaviour is exactly as it should be.
In this case the returned coordinates should be exactly in the center of the text.  

Here are some screenshots I took while testing. 

With the existing changes: 
https://drive.google.com/file/d/1OMjYDEh03vS4Rflgldxu5ruN8qzsN7uK/view?usp=sharing
https://drive.google.com/file/d/12aaEzku-waJbM_BHLTA3g5oBf_O4pKG9/view?usp=sharing

With suggested changes: 
https://drive.google.com/file/d/1SPbV4IUhP4bC6CRjCtLW-iRux71CgdN_/view?usp=sharing
https://drive.google.com/file/d/1fu0pbYGCF2lXVn19H2psCJqQpdOFv_eh/view?usp=sharing

**Comment by ritiksp2411** [comment]
Thanks for the insight and for helping me out. I will change the code as suggested and update the PR soon.

File: catroid/src/main/java/org/catrobat/catroid/utils/TextBlockUtil.java [hunk 3]
```diff
@@ -102,13 +102,13 @@ public static Point getCenterCoordinates(int index) {
...
+					float relativeX = textBlockBounds.exactCenterX() / imageWidth;
+					relativeX = invertAxis ? 1 - relativeX : relativeX;
 					return coordinatesFromRelativePosition(
-							1 - textBlockBounds.exactCenterY() / imageHeight,
-							SCREEN_WIDTH / aspectRatio,
-							relativeY,
-							(float) SCREEN_WIDTH
+							relativeX,
+							SCREEN_WIDTH,
+							1 - textBlockBounds.exactCenterY() / imageWidth,
```
Comment (qlin93) [inline catroid/src/main/java/org/catrobat/catroid/utils/TextBlockUtil.java:110]:
```suggestion
							1 - textBlockBounds.exactCenterY() / imageHeight,
```

**Review by qlin93 — CHANGES_REQUESTED** [review]
Thanks just a small error, please change imageWidth to imageHeight in line 110.

**Review by qlin93 — APPROVED** [review]
Thanks!

**Comment by ritiksp2411** [comment]
Thanks for helping,
Learned a lot from the issue

**Comment by wslany** [comment]
Thanks a lot!!

**Pull request merged**